### PR TITLE
Make openFile exception safe wrt #8

### PR DESCRIPTION
**Pull request by hasufell**
@Mistuke @treeowl 

Related: https://github.com/hasufell/file-io/issues/8

### CODE REVIEW

File: System/File/OsPath.hs [hunk 1]
```diff
@@ -98,3 +115,38 @@ openFile (OsString fp) = P.openFile fp
...
+      FileHandle fp m -> (fp, m)
+      DuplexHandle fp _ write_m -> (fp, write_m)
+
+withOpenFile' :: (OsPath -> IOMode -> IO Handle) -> OsPath -> IOMode -> (Handle -> IO r) -> IO r
+withOpenFile' acquire fp iomode action = mask $ \restore -> do
+  hndl <- acquire fp iomode
+  addHandleFinalizer hndl handleFinalizer
+  r <- restore (action hndl) `onException` hClose hndl
+  hClose hndl
+  pure r
```
Comment (hasufell) [inline System/File/OsPath.hs:151]: This is the juice. It's supposed to simulate:

* https://hackage.haskell.org/package/base-4.17.0.0/docs/src/GHC.IO.Handle.FD.html#withOpenFile%27

**Comment by hasufell** [comment]
ping @treeowl

File: posix/System/File/Platform.hs [hunk 2]
```diff
@@ -37,7 +38,7 @@ openExistingFile fp iomode = fdToHandle_ iomode fp =<< case iomode of
   df = defaultFileFlags { noctty = True, nonBlock = True, creat = Nothing }
 
 fdToHandle_ :: IOMode -> PosixPath -> Fd -> IO Handle
-fdToHandle_ iomode fp (Fd fd) = do
+fdToHandle_ iomode fp (Fd fd) = (`onException` closeFd (Fd fd)) $ do
```
Comment (hasufell) [inline posix/System/File/Platform.hs:41]: from what I understand, `fdToHandle'` does not close the fd on failure, so I guess we need the `onException` here?

File: windows/System/File/Platform.hsc [hunk 3]
```diff
@@ -43,11 +43,11 @@ openFile fp iomode = bracketOnError
     toHandle
  where
 #if defined(__IO_MANAGER_WINIO__)
-  toHandle h = do
+  toHandle h = (`onException` Win32.closeHandle h) $ do
```
Comment (hasufell) [inline windows/System/File/Platform.hsc:46]: Not 100% sure if we want `closeHandle` here, but I can't see any such attempt in `hANDLEToHandle` etc. So I guess we do that here.

File: System/File/OsPath.hs [hunk 4]
```diff
@@ -118,11 +117,48 @@ appendFile' fp contents = withFile fp AppendMode (`BS.hPut` contents)
...
+withOpenFile' :: OsPath -> IOMode -> Bool -> Bool -> (Handle -> IO r) -> Bool -> IO r
+withOpenFile' (OsString fp) iomode binary existing action close_finally = mask $ \restore -> do
+  hndl <- if existing
+          then P.openExistingFile fp iomode
+          else P.openFile fp iomode
+  addHandleFinalizer hndl handleFinalizer
+  when binary $ hSetBinaryMode hndl True
+  r <- restore (action hndl) `onException` hClose hndl
+  when close_finally $ hClose hndl
+  pure r
```
Comment (hasufell) [inline System/File/OsPath.hs:161]: This is the juice.

**Review by Bodigrim — APPROVED** [review]
Checked it locally on macOS, seems good.

**Comment by hasufell** [comment]
@treeowl i want to merge and release it this week. Do you have time to review?

**Comment by treeowl** [comment]
Is this just a translation of what I did in `base`? If so, sounds good. I don't have everything paged in, I was out of the country this week, and I have a lot of things I need to do tomorrow, so I don't think I can handle a full scale review.

**Comment by hasufell** [comment]
> Is this just a translation of what I did in base? 

Yes, the major relevant parts are:

https://github.com/hasufell/file-io/blob/01c64dfe304d4006a8215fa908ac675a9ac95390/System/File/OsPath.hs#L131-L161

and then `readFile`, `openFile` etc. are defined in terms of `withOpenFile'`.

**Pull request merged**